### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by yuyuranium**
I can reproduce the problem encountered in #6  so I wrote some test to test both BRAM and SuperBRAM. However, no matter how I tried, the result seems to be the same: The initial values in both BRAM and SuperBRAM are all shifted by one word. See the program output below:

image

The output is generated by the function `dump_bram` in `main.c`.

I also did some test on writing data to brams and reading it back by the codes I commented below. The result was weird:

image

The main program should write `0x87` to word offset 0, 5, and 8 of both brams (But it's offset 1, 6, and 9 that become `0x87`). From the picture we can tell that all writing operations are shifted by one word also. Besides, the offset 0 of both brams were also be written.

BTW, the good news is that `codegen` function is working fine.

### CODE REVIEW

File: src/main.c [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	printf("===== Initial BRAM values =====\n");
+	dump_bram();
+	//exec(XPAR_CONTROLLER_0_S00_AXI_BASEADDR,
+		// codegen(0, 0b0000101, 0b10001, 3 + 1, 2 + 1, 0 + 1)); // alumode = 0000, opmode = 0000101, inmode = 10001, bram1wAddr = 4*3, bram1rAddr = 4*2, bram0rAddr = 4*0;
+	Xil_Out32(XPAR_AXI_BRAM_CTRL_1_S_AXI_BASEADDR, 0x87);
+	Xil_Out32(XPAR_AXI_BRAM_CTRL_1_S_AXI_BASEADDR + 20, 0x87);
+	Xil_Out32(XPAR_AXI_BRAM_CTRL_0_S_AXI_BASEADDR, 0x87);
+	Xil_Out32(XPAR_AXI_BRAM_CTRL_0_S_AXI_BASEADDR + 20, 0x87);
+	Xil_Out32(XPAR_AXI_BRAM_CTRL_1_S_AXI_BASEADDR + 32, 0x87);
+	Xil_Out32(XPAR_AXI_BRAM_CTRL_0_S_AXI_BASEADDR + 32, 0x87);
```
Comment (yuyuranium) [inline src/main.c]: I used these codes to test reading/writing the brams.

**Comment by yuyuranium** [comment]
## Update
I modified the `dump_bram` function to read a single word multiple times and output the final result. Now it looks all good!

- Initial value
image

- After writing `0x87` to 0, 5, 8 word offsets
image

File: src/main.c [hunk 2]
```diff
@@ -13,9 +13,12 @@
...
 {
 	for (int i = 0; i < 33; ++i) {
-		printf("BRAM[%d] = %lx \t \t \t SuperBRAM[%d] = %lx\n",
-				i, Xil_In32(XPAR_AXI_BRAM_CTRL_0_S_AXI_BASEADDR + 4 * i),
-				i, Xil_In32(XPAR_AXI_BRAM_CTRL_1_S_AXI_BASEADDR + 4 * i));
+		u32 a, b;
+		for (int j = 0; j < 10; ++j) {
+			a = Xil_In32(XPAR_AXI_BRAM_CTRL_0_S_AXI_BASEADDR + 4 * i);
+			b = Xil_In32(XPAR_AXI_BRAM_CTRL_1_S_AXI_BASEADDR + 4 * i);
+		}
```
Comment (yuyuranium) [inline src/main.c]: I used this for loop to read the same entry multiple times (exactly 10 times) to get a steady result.

**Comment by yuyuranium** [comment]
## Update
Good news! Thanks to #9, the brams work perfectly!

**Comment by WillyChennnn** [comment]
Finally,
thanks for testing！

**Comment by yuyuranium** [comment]
Bad news is that the computation seems not to be carried out. I am checking if the output `P` from DSP has value or not by instanciate an AXI_GPIO so that the PS can read the value of `P` and print it out on the terminal. QQ

**Comment by yuyuranium** [comment]
I have test the value of `P` via AXI_GPIO. However, the value before and after executing the command is always 0.

image

File: IPs/controller_1.0/drivers/controller_v1_0/src/controller.c [hunk 3]
```diff
@@ -10,9 +10,16 @@ void exec(UINTPTR baseAddr, u32 ins)
...
+#include "xparameters.h"
+    printf("Before: %lx\n", Xil_In32(XPAR_AXI_GPIO_0_BASEADDR));
 
     // Busy waiting for execution
     while (CONTROLLER_mReadReg(baseAddr, 0));
+    printf("After: %lx\n", Xil_In32(XPAR_AXI_GPIO_0_BASEADDR));
+    printf("After: %lx\n", Xil_In32(XPAR_AXI_GPIO_0_BASEADDR));
+    printf("After: %lx\n", Xil_In32(XPAR_AXI_GPIO_0_BASEADDR));
+    printf("After: %lx\n", Xil_In32(XPAR_AXI_GPIO_0_BASEADDR));
+    printf("After: %lx\n", Xil_In32(XPAR_AXI_GPIO_0_BASEADDR));
```
Comment (yuyuranium) [inline IPs/controller_1.0/drivers/controller_v1_0/src/controller.c]: I used these code to read the value from AXI_GPIO, which is added to the block design by me to read the output of DSP directly.

**Comment by yuyuranium** [comment]
### Update
Good news! The system seems to work. I revise the controller and how it communicates with PS. Now PS has to use AXI GPIO to invoke the execution of an instruction by
1. Set the instruction to controller 
2. Write enable signal to controller
3. Wait for valid signal from controller
4. Reset enable signal to controller

Here is the result of executing the first instruction:
- Before
image

- After
image

We can see that after executiong SuperBRAM[3] is written with correct value.

**Comment by yuyuranium** [comment]
@WillyChennnn would you help me translate the pseudo code into a series of `codegen` function calls?

For example,
```
BRAM1[3] <= BRAM0[0] * BRAM1[2];
```
becomes
```
codegen(0b0000, 0b0000101, 0b10001, 3, 2, 0)
```

You may write the result directly in the comment section, thanks!

**Comment by WillyChennnn** [comment]
Ok, but may I complete this tomorrow?
I will have an interview for internship in tomorrow morning, so I want to spend some time preparing it😥

**Comment by yuyuranium** [comment]
Oh! I noticed that you have done so in the comments in `main.c`!

**Comment by yuyuranium** [comment]
### Update
Thanks to @WillyChennnn's dedication, our system finally works perfectly. Here is the execution result, which is totally identical to TA's answer.

image
image

I think this PR is ready for review. Please let me know if you have comments. Thanks a lot!

**Comment by WillyChennnn** [comment]
Congratulations
Thanks for big hardworking!

**Comment by yuyuranium** [comment]
BTW, I added the vitis project and removed the original controller IP. Also, here is the block diagram:

image

**Review by japoka410666 — APPROVED** [review]
Thanks for your hrad working anb testing. It looks great !!

**Pull request merged**